IpmiFeaturePkg/GenericIpmi/Dxe/DxeGenericIpmi.c

```diff
@@ -1,51 +1,447 @@
+
+  //
+  // Set up a loop to retry for up to PcdIpmiBmcReadyDelayTimer seconds. Calculate retries not timeout
+  // so that in case KCS is not enabled and IpmiSendCommand() returns
+  // immediately we will not wait all the PcdIpmiBmcReadyDelayTimer seconds.
+  //
+  Retries = PcdGet8 (PcdIpmiBmcReadyDelayTimer);
+  //
+  // Get the device ID information for the BMC.
+  //
+  DataSize = sizeof (IpmiInstance->TempData);
+  DEBUG ((DEBUG_INFO, "[IPMI] Getting BMC Device ID. DataSize: 0x%x Retries: %d\n", DataSize, Retries));
+  while (EFI_ERROR (
+           Status = IpmiSendCommand (
+                      &IpmiInstance->IpmiTransport,
+                      IPMI_NETFN_APP,
+                      0,
+                      IPMI_APP_GET_DEVICE_ID,
+                      NULL,
+                      0,
+                      IpmiInstance->TempData,
+                      &DataSize
+                      )
+           )
+         )
+  {
+    DEBUG ((
+      DEBUG_ERROR,
+      "[IPMI] BMC does not respond by Get BMC DID (status: %r), %d retries left.\n",
+      Status,
+      Retries
+      ));
+
+    if (Retries-- == 0) {
+      IpmiInstance->BmcStatus = BMC_HARDFAIL;
+      return Status;
+    }
+
+    //
+    // Handle the case that BMC FW still not enable KCS channel after AC cycle. just stall 1 second
+    //
+    MicroSecondDelay (1*1000*1000);
+  }
+
+  pBmcInfo = (SM_CTRL_INFO *)&IpmiInstance->TempData[0];
+  DEBUG ((DEBUG_INFO, "[IPMI] BMC Device ID: 0x%02X, firmware version: %d.%02X UpdateMode:%x\n", pBmcInfo->DeviceId, pBmcInfo->MajorFirmwareRev, pBmcInfo->MinorFirmwareRev, pBmcInfo->UpdateMode));
+  //
+  // In OpenBMC, UpdateMode: the bit 7 of byte 4 in get device id command is used for the BMC status:
+  // 0 means BMC is ready, 1 means BMC is not ready.
+  // At the very beginning of BMC power on, the status is 1 means BMC is in booting process and not ready. It is not the flag for force update mode.
+  //
+  if (pBmcInfo->UpdateMode == BMC_READY) {
+    mIpmiInstance->BmcStatus = BMC_OK;
+    return EFI_SUCCESS;
+  } else {
+    Status = IpmiSendCommand (
+               &IpmiInstance->IpmiTransport,
+               IPMI_NETFN_FIRMWARE,
+               0,
+               IPMI_GET_BMC_EXECUTION_CONTEXT,
+               NULL,
+               0,
+               IpmiInstance->TempData,
+               &DataSize
+               );
+
+    pBmcExecContext = (IPMI_MSG_GET_BMC_EXEC_RSP *)&IpmiInstance->TempData[0];
+    DEBUG ((DEBUG_INFO, "[IPMI] Operational status of BMC: 0x%x\n", pBmcExecContext->CurrentExecutionContext));
+    if ((pBmcExecContext->CurrentExecutionContext == IPMI_BMC_IN_FORCED_UPDATE_MODE) &&
+        !EFI_ERROR (Status))
+    {
+      DEBUG ((DEBUG_WARN, "[IPMI] BMC in Forced Update mode, skip waiting for BMC_READY.\n"));
+      IpmiInstance->BmcStatus = BMC_UPDATE_IN_PROGRESS;
+    } else {
+      //
+      // Updatemode = 1 mean BMC is not ready, continue waiting.
+      //
+      while (Retries-- != 0) {

```

Comment:
layman question: do you need to reset this counter for this operation?

Reply:
Yeah probably. will do